compiler/parser/RdrHsSyn.hs

```diff
@@ -2611,14 +2611,23 @@ mkTemplateClassMethod ::
   -> LHsExpr GhcPs               -- ^ function body
   -> Maybe (LHsLocalBinds GhcPs) -- ^ local binds
   -> LHsBind GhcPs               -- ^ function binding
-mkTemplateClassMethod rawMethodName args body mBinds = do
+mkTemplateClassMethod rawMethodName =
   let fullMethodName = noLoc $ mkRdrUnqual $ mkVarOcc rawMethodName
-      ctx = matchContext fullMethodName
+  in mkBind fullMethodName
+
+mkBind ::
+     Located RdrName             -- ^ bound name
+  -> [Pat GhcPs]                 -- ^ argument patterns
+  -> LHsExpr GhcPs               -- ^ RHS
+  -> Maybe (LHsLocalBinds GhcPs) -- ^ local binds
+  -> LHsBind GhcPs               -- ^ function binding
+mkBind boundName args body mBinds = do
+  let ctx = matchContext boundName
       loc = getLoc body
       binds = fromMaybe (noLoc emptyLocalBinds) mBinds
       match = matchWithBinds ctx args loc body binds
       match_group = matchGroup loc match
-  L loc $ funBind fullMethodName match_group
+  L loc $ funBind boundName match_group
```

Comment:
I think _this_ was the important change. Adding the location on `topLevelFuncSig` and `topLevelFuncDef` alone didn't make a difference. IIUC, the reason for this is that ghcide looks for value definitions through their `LHsBind`s, not through the `LHsDecl`s constructed with `SigD` or `ValD`